<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Bulk Tidy Transcriptomics - analysis of bulk RNA sequencing data with R tidy principles • zhejiang2020</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Bulk Tidy Transcriptomics - analysis of bulk RNA sequencing data with R tidy principles">
<meta property="og:description" content="zhejiang2020">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">zhejiang2020</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Bulk analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">R base</li>
    <li>
      <a href="../articles/limmaWorkflow.html">English</a>
    </li>
    <li>
      <a href="../articles/limmaWorkflow_chinese.html">Chinese</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Tidy transcriptomics</li>
    <li>
      <a href="../articles/tidy_limmaWorkflow.html">English</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Single-cell analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">R base</li>
    <li>
      <a href="../articles/sc_analysis.html">Chinese</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Tidy transcriptomics</li>
    <li>
      <a href="../articles/tidy_single_cell.html">English</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/zhejiang2020">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/repository/docker/stemangiola/zhejiang2020">
    <span class="fa fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tidy_limmaWorkflow_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="tidy_limmaWorkflow_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="tidy_limmaWorkflow_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Bulk Tidy Transcriptomics - analysis of bulk RNA sequencing data with R tidy principles</h1>
                        <h4 class="author">Stefano Mangiola</h4>
            <address class="author_afil">
      The Walter and Eliza Hall Institute of Medical Research, 1G Royal Parade, Parkville, VIC 3052, Melbourne, Australia; Department of Medical Biology, The University of Melbourne, Parkville, VIC 3010, Melbourne, Australia<br><h4 class="author">Maria Doyle</h4>
            <address class="author_afil">
      Peter MacCallum Cancer Centre, 305 Grattan Street, Parkville, Melbourne, Victoria, Australia<br><h4 class="date">11 November 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/zhejiang2020/blob/master/vignettes/tidy_limmaWorkflow.Rmd"><code>vignettes/tidy_limmaWorkflow.Rmd</code></a></small>
      <div class="hidden name"><code>tidy_limmaWorkflow.Rmd</code></div>

    </address>
</address>
</div>

    
    
<p><img src="tidybulk_logo.png" width="100"></p>
<div id="abstract" class="section level1">
<h1 class="hasAnchor">
<a href="#abstract" class="anchor"></a>Abstract</h1>
<p>This is the <code>tidy</code> version of the material in the session <code>RNA-seq analysis is easy as 1-2-3 with limma, Glimma and edgeR</code> of this workshop</p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Measuring gene expression on a genome-wide scale has become common practice over the last two decades or so, with microarrays predominantly used pre-2008. With the advent of next generation sequencing technology in 2008, an increasing number of scientists use this technology to measure and understand changes in gene expression in often complex systems. As sequencing costs have decreased, using RNA sequencing to simultaneously measure the expression of tens of thousands of genes for multiple samples has never been easier. The cost of these experiments has now moved from generating the data to storing and analysing it.</p>
<p>There are many steps involved in analysing an RNA sequencing dataset. Sequenced reads are aligned to a reference genome, then the number of reads mapped to each gene can be counted. This results in a table of counts, which is what we perform statistical analyses on in R. While mapping and counting are important and necessary tasks, today we will be starting from the count data and showing how differential expression analysis can be performed in a friendly way using the Bioconductor package, tidybulk.</p>
</div>
<div id="set-up" class="section level1">
<h1 class="hasAnchor">
<a href="#set-up" class="anchor"></a>Set-up</h1>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stemangiola.github.io/zhejiang2020/">zhejiang2020</a></span><span class="op">)</span>

<span class="co"># tidyverse core packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>

<span class="co"># tidyverse-friendly packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://github.com/slowkow/ggrepel">ggrepel</a></span><span class="op">)</span>

<span class="co">#library(tidyHeatmap)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidybulk</span><span class="op">)</span></code></pre></div>
<p>Plot settings. Set the colours and theme we will use for our plots.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Use colourblind-friendly colours</span>
<span class="va">friendly_cols</span> <span class="op">&lt;-</span> <span class="fu">dittoSeq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html">dittoColors</a></span><span class="op">(</span><span class="op">)</span>
<span class="co"># Set theme</span>
<span class="va">custom_theme</span> <span class="op">&lt;-</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">friendly_cols</span><span class="op">)</span>,
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
      <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
        panel.border <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.line <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span><span class="op">)</span>,
        text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span>,
        legend.position <span class="op">=</span> <span class="st">"bottom"</span>,
        strip.background <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
        axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>
      <span class="op">)</span>
  <span class="op">)</span></code></pre></div>
</div>
<div id="data-packaging" class="section level1">
<h1 class="hasAnchor">
<a href="#data-packaging" class="anchor"></a>Data packaging</h1>
<div id="organising-information-within-a-tibble-user-friendly-table" class="section level2">
<h2 class="hasAnchor">
<a href="#organising-information-within-a-tibble-user-friendly-table" class="anchor"></a>Organising information within a <code>tibble</code> (user-friendly table)</h2>
<p>We can create a tibble with our gene-transcript abundance and sample annotation information</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts</span> <span class="op">=</span> 
  
  <span class="co"># Create the counts table</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span>
  
    <span class="co"># Transcript abundance</span>
    <span class="fu">zhejiang2020</span><span class="fu">::</span><span class="va"><a href="../reference/dge_list.html">dge_list</a></span><span class="op">$</span><span class="va">counts</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames<span class="op">=</span><span class="st">"entrez"</span><span class="op">)</span> <span class="op">%&gt;%</span> 
      <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span><span class="op">-</span><span class="va">entrez</span>, names_to<span class="op">=</span><span class="st">"sample"</span>, values_to<span class="op">=</span><span class="st">"count"</span><span class="op">)</span>,
    
    <span class="co"># Sample annotation</span>
     <span class="fu">zhejiang2020</span><span class="fu">::</span><span class="va"><a href="../reference/dge_list.html">dge_list</a></span><span class="op">$</span><span class="va">samples</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames<span class="op">=</span><span class="st">"sample"</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="op">-</span><span class="va">lib.size</span>, <span class="op">-</span><span class="va">norm.factors</span>, <span class="op">-</span><span class="va">files</span><span class="op">)</span>
    
  <span class="op">)</span> <span class="op">%&gt;%</span>
    
  <span class="co"># Add gene symbols</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu">AnnotationDbi</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/AnnotationDb-class.html">mapIds</a></span><span class="op">(</span>
      <span class="fu">org.Mm.eg.db</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/org.Mm.eg.db/man/org.Mm.egBASE.html">org.Mm.eg.db</a></span>, 
      keys <span class="op">=</span> <span class="va">entrez</span>, 
      keytype <span class="op">=</span> <span class="st">"ENTREZID"</span>, 
      column<span class="op">=</span><span class="st">"SYMBOL"</span>, 
      multiVals <span class="op">=</span> <span class="st">"first"</span>
    <span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Filter for empty symbols</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Memorise key column names for streamlined analyses using `tidybulk`</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/tidybulk-methods.html">tidybulk</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">symbol</span>, <span class="va">count</span><span class="op">)</span></code></pre></div>
<pre><code>## Joining, by = "sample"</code></pre>
<pre><code>## </code></pre>
<pre><code>## 'select()' returned 1:1 mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts</span></code></pre></div>
<pre><code>## # A tibble: 233,136 x 6
##    entrez    sample    count group lane  symbol 
##    &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt; &lt;fct&gt; &lt;fct&gt; &lt;chr&gt;  
##  1 497097    10_6_5_11     1 LP    L004  Xkr4   
##  2 497097    9_6_5_11      2 ML    L004  Xkr4   
##  3 497097    purep53     342 Basal L004  Xkr4   
##  4 497097    JMS8-2      526 Basal L006  Xkr4   
##  5 497097    JMS8-3        3 ML    L006  Xkr4   
##  6 497097    JMS8-4        3 LP    L006  Xkr4   
##  7 497097    JMS8-5      535 Basal L006  Xkr4   
##  8 497097    JMS9-P7c      2 ML    L008  Xkr4   
##  9 497097    JMS9-P8c      0 LP    L008  Xkr4   
## 10 100503874 10_6_5_11     0 LP    L004  Gm19938
## # … with 233,126 more rows</code></pre>
</div>
</div>
<div id="aggregate-duplicate-gene-symbols" class="section level1">
<h1 class="hasAnchor">
<a href="#aggregate-duplicate-gene-symbols" class="anchor"></a>Aggregate duplicate gene symbols</h1>
<p>Our gene annotation contains 28 genes that map to multiple chromosomes, in this case we will combine all chromosome information from the multi-mapped genes</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_aggregated</span> <span class="op">=</span> 
  <span class="va">counts</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/aggregate_duplicates-methods.html">aggregate_duplicates</a></span><span class="op">(</span>aggregation_function <span class="op">=</span> <span class="va">median</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning in aggregate_duplicated_transcripts_bulk(.data, .sample = !!.sample, : tidybulk
## says: for aggregation, factors and logical columns were converted to character</code></pre>
<pre><code>## Converted to characters</code></pre>
<pre><code>## factorfactor</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_aggregated</span></code></pre></div>
<pre><code>## # A tibble: 233,136 x 7
##    entrez    sample    count group lane  symbol  `merged transcripts`
##    &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;                  &lt;dbl&gt;
##  1 497097    10_6_5_11     1 LP    L004  Xkr4                       1
##  2 497097    9_6_5_11      2 ML    L004  Xkr4                       1
##  3 497097    purep53     342 Basal L004  Xkr4                       1
##  4 497097    JMS8-2      526 Basal L006  Xkr4                       1
##  5 497097    JMS8-3        3 ML    L006  Xkr4                       1
##  6 497097    JMS8-4        3 LP    L006  Xkr4                       1
##  7 497097    JMS8-5      535 Basal L006  Xkr4                       1
##  8 497097    JMS9-P7c      2 ML    L008  Xkr4                       1
##  9 497097    JMS9-P8c      0 LP    L008  Xkr4                       1
## 10 100503874 10_6_5_11     0 LP    L004  Gm19938                    1
## # … with 233,126 more rows</code></pre>
</div>
<div id="data-pre-processing" class="section level1">
<h1 class="hasAnchor">
<a href="#data-pre-processing" class="anchor"></a>Data pre-processing</h1>
<div id="identifying-genes-that-are-lowly-expressed" class="section level2">
<h2 class="hasAnchor">
<a href="#identifying-genes-that-are-lowly-expressed" class="anchor"></a>Identifying genes that are lowly expressed</h2>
<p>As before, identifying lowly transcribed genes is necessary for several downstream analyses. We can specify the <code>factor of interest</code> for a more informed filtering. This function uses the edgeR utility <code>filterByExpr</code>.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_abundant</span> <span class="op">=</span> 
  <span class="va">counts_aggregated</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/identify_abundant-methods.html">identify_abundant</a></span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">group</span><span class="op">)</span>

<span class="va">counts_abundant</span></code></pre></div>
<pre><code>## # A tibble: 233,136 x 8
##    entrez    sample    count group lane  symbol  `merged transcripts` .abundant
##    &lt;chr&gt;     &lt;chr&gt;     &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;                  &lt;dbl&gt; &lt;lgl&gt;    
##  1 497097    10_6_5_11     1 LP    L004  Xkr4                       1 TRUE     
##  2 497097    9_6_5_11      2 ML    L004  Xkr4                       1 TRUE     
##  3 497097    purep53     342 Basal L004  Xkr4                       1 TRUE     
##  4 497097    JMS8-2      526 Basal L006  Xkr4                       1 TRUE     
##  5 497097    JMS8-3        3 ML    L006  Xkr4                       1 TRUE     
##  6 497097    JMS8-4        3 LP    L006  Xkr4                       1 TRUE     
##  7 497097    JMS8-5      535 Basal L006  Xkr4                       1 TRUE     
##  8 497097    JMS9-P7c      2 ML    L008  Xkr4                       1 TRUE     
##  9 497097    JMS9-P8c      0 LP    L008  Xkr4                       1 TRUE     
## 10 100503874 10_6_5_11     0 LP    L004  Gm19938                    1 FALSE    
## # … with 233,126 more rows</code></pre>
</div>
</div>
<div id="scaling-data-for-visualisation" class="section level1">
<h1 class="hasAnchor">
<a href="#scaling-data-for-visualisation" class="anchor"></a>Scaling data for visualisation</h1>
<p>We can compensate for technical differences in sequencing depth, scaling the data (also called normalisation). By default the <code>TMM</code> <span class="citation">(Robinson and Oshlack 2010)</span> method is used. The scaling will be calculated on the highly-transcribed genes and applied on all genes.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scaled</span> <span class="op">=</span> 
  <span class="va">counts_abundant</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/scale_abundance-methods.html">scale_abundance</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">counts_scaled</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">symbol</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span><span class="op">(</span><span class="st">"count"</span><span class="op">)</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 233,136 x 11
##    sample symbol count count_scaled entrez group lane  `merged transcr… .abundant   TMM
##    &lt;chr&gt;  &lt;chr&gt;  &lt;dbl&gt;        &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;            &lt;dbl&gt; &lt;lgl&gt;     &lt;dbl&gt;
##  1 10_6_… Xkr4       1         2.59 497097 LP    L004                 1 TRUE      0.891
##  2 9_6_5… Xkr4       2         4.14 497097 ML    L004                 1 TRUE      1.04 
##  3 purep… Xkr4     342       421.   497097 Basal L004                 1 TRUE      1.08 
##  4 JMS8-2 Xkr4     526       762.   497097 Basal L006                 1 TRUE      1.02 
##  5 JMS8-3 Xkr4       3         2.91 497097 ML    L006                 1 TRUE      1.03 
##  6 JMS8-4 Xkr4       3         4.10 497097 LP    L006                 1 TRUE      0.916
##  7 JMS8-5 Xkr4     535       729.   497097 Basal L006                 1 TRUE      1.01 
##  8 JMS9-… Xkr4       2         6.60 497097 ML    L008                 1 TRUE      1.08 
##  9 JMS9-… Xkr4       0         0    497097 LP    L008                 1 TRUE      0.959
## 10 10_6_… Gm199…     0         0    10050… LP    L004                 1 FALSE     0.891
## # … with 233,126 more rows, and 1 more variable: multiplier &lt;dbl&gt;</code></pre>
<p>We can reproduce the log-transcript-abundance density of unfiltered and filtered data (seen in the previous session of the workshop) using tidyverse tools</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/bind.html">bind_rows</a></span><span class="op">(</span>
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>label <span class="op">=</span> <span class="st">"1.Unfiltered"</span><span class="op">)</span>,
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.abundant</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>label <span class="op">=</span> <span class="st">"2.Filtered"</span><span class="op">)</span>
<span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">count_scaled</span> <span class="op">+</span><span class="fl">1</span>, color <span class="op">=</span> <span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> 
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">label</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidy_limmaWorkflow_files/figure-html/unnamed-chunk-4-1.png" width="700"></p>
<p>We can reproduce the log-transcript-abundance density of unscaled and scaled data (seen in the previous session of the workshop) using tidyverse tools</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.abundant</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># We reshape the data in order to build a faceted plot</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html">contains</a></span><span class="op">(</span><span class="st">"count"</span><span class="op">)</span>, names_to<span class="op">=</span><span class="st">"processing"</span>, values_to<span class="op">=</span><span class="st">"value"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Build the plot</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">value</span> <span class="op">+</span> <span class="fl">1</span>, fill<span class="op">=</span><span class="va">sample</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">processing</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_y_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<p><img src="tidy_limmaWorkflow_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<div id="dimensionality-reduction" class="section level2">
<h2 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>Dimensionality reduction</h2>
<p>As previously shown, we can perform dimensionality reduction to further explore our data. The <code>reduce_dimensions</code> function, will perform calculations only on highly-transcribed genes (i.e. .abundant == TRUE)</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/reduce_dimensions-methods.html">reduce_dimensions</a></span><span class="op">(</span>method <span class="op">=</span> <span class="st">"MDS"</span>, action<span class="op">=</span><span class="st">"get"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># We reshape the data in order to build a faceted plot</span>
  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/pivot_longer.html">pivot_longer</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">group</span>, <span class="va">lane</span><span class="op">)</span>, names_to<span class="op">=</span><span class="st">"annotation"</span>, values_to<span class="op">=</span><span class="st">"value"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Build the plot</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">Dim1</span>, <span class="va">Dim2</span>, color<span class="op">=</span><span class="va">value</span>, label<span class="op">=</span><span class="va">value</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html">geom_text</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html">facet_wrap</a></span><span class="op">(</span><span class="op">~</span><span class="va">annotation</span><span class="op">)</span> <span class="op">+</span>
  <span class="va">custom_theme</span></code></pre></div>
<pre><code>## tidybulk says: to access the raw results do `attr(..., "internals")$MDS`</code></pre>
<p><img src="tidy_limmaWorkflow_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
</div>
</div>
<div id="differential-expression-analysis" class="section level1">
<h1 class="hasAnchor">
<a href="#differential-expression-analysis" class="anchor"></a>Differential expression analysis</h1>
<p>We can replicate the differential expression analyses using <code>tidybulk</code></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html">model.matrix</a></span><span class="op">(</span>
  <span class="op">~</span><span class="fl">0</span><span class="op">+</span><span class="va">group</span><span class="op">+</span><span class="va">lane</span>,
  data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/pivot_sample-methods.html">pivot_sample</a></span><span class="op">(</span><span class="va">counts_scaled</span><span class="op">)</span>
<span class="op">)</span></code></pre></div>
<pre><code>##   groupBasal groupLP groupML laneL006 laneL008
## 1          0       1       0        0        0
## 2          0       0       1        0        0
## 3          1       0       0        0        0
## 4          1       0       0        1        0
## 5          0       0       1        1        0
## 6          0       1       0        1        0
## 7          1       0       0        1        0
## 8          0       0       1        0        1
## 9          0       1       0        0        1
## attr(,"assign")
## [1] 1 1 1 2 2
## attr(,"contrasts")
## attr(,"contrasts")$group
## [1] "contr.treatment"
## 
## attr(,"contrasts")$lane
## [1] "contr.treatment"</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_test</span> <span class="op">=</span> 
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidybulk/man/test_differential_abundance-methods.html">test_differential_abundance</a></span><span class="op">(</span>
    .formula <span class="op">=</span> <span class="op">~</span><span class="fl">0</span><span class="op">+</span><span class="va">group</span><span class="op">+</span><span class="va">lane</span>,
    .contrasts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"groupBasal-groupLP"</span>, <span class="st">"groupBasal - groupML"</span>, <span class="st">"groupLP - groupML"</span><span class="op">)</span>,
    method <span class="op">=</span> <span class="st">"limma_voom"</span>,
    action<span class="op">=</span><span class="st">"get"</span>
  <span class="op">)</span></code></pre></div>
<pre><code>## =====================================
## tidybulk says: All testing methods use raw counts, irrespective of if scale_abundance 
## or adjust_abundance have been calculated. Therefore, it is essential to add covariates 
## such as batch effects (if applicable) in the formula.
## =====================================</code></pre>
<pre><code>## tidybulk says: The design column names are "groupBasal, groupLP, groupML, laneL006, laneL008"</code></pre>
<pre><code>## tidybulk says: to access the raw results (fitted GLM) do `attr(..., "internals")$voom`</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_test</span></code></pre></div>
<pre><code>## # A tibble: 25,904 x 22
##    symbol entrez `merged transcr… .abundant `logFC___groupB… `logFC___groupB…
##    &lt;chr&gt;  &lt;chr&gt;             &lt;dbl&gt; &lt;lgl&gt;                &lt;dbl&gt;            &lt;dbl&gt;
##  1 Xkr4   497097                1 TRUE                 7.44           6.94   
##  2 Gm199… 10050…                1 FALSE               NA             NA      
##  3 Gm105… 10003…                1 FALSE               NA             NA      
##  4 Rp1    19888                 1 FALSE               NA             NA      
##  5 Sox17  20671                 1 TRUE                 2.93           2.30   
##  6 Mrpl15 27395                 1 TRUE                 0.466          0.00912
##  7 Lypla1 18777                 1 TRUE                 0.275         -0.395  
##  8 Gm198… 10050…                1 FALSE               NA             NA      
##  9 Tcea1  21399                 1 TRUE                 0.354          0.380  
## 10 Rgs20  58175                 1 TRUE                -5.50          -5.05   
## # … with 25,894 more rows, and 16 more variables: `logFC___groupLP - groupML` &lt;dbl&gt;,
## #   `AveExpr___groupBasal-groupLP` &lt;dbl&gt;, `AveExpr___groupBasal - groupML` &lt;dbl&gt;,
## #   `AveExpr___groupLP - groupML` &lt;dbl&gt;, `t___groupBasal-groupLP` &lt;dbl&gt;, `t___groupBasal
## #   - groupML` &lt;dbl&gt;, `t___groupLP - groupML` &lt;dbl&gt;,
## #   `P.Value___groupBasal-groupLP` &lt;dbl&gt;, `P.Value___groupBasal - groupML` &lt;dbl&gt;,
## #   `P.Value___groupLP - groupML` &lt;dbl&gt;, `adj.P.Val___groupBasal-groupLP` &lt;dbl&gt;,
## #   `adj.P.Val___groupBasal - groupML` &lt;dbl&gt;, `adj.P.Val___groupLP - groupML` &lt;dbl&gt;,
## #   `B___groupBasal-groupLP` &lt;dbl&gt;, `B___groupBasal - groupML` &lt;dbl&gt;, `B___groupLP -
## #   groupML` &lt;dbl&gt;</code></pre>
<p>We can reproduce the fitted means (x-axis) and variances (y-axis) relationship of each gene, using the raw results from limma-voom.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_test</span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/r/base/attr.html">attr</a></span><span class="op">(</span><span class="st">"internals"</span><span class="op">)</span> <span class="op">%$%</span> 
  <span class="va">voom</span> <span class="op">%&gt;%</span> 
  <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html">eBayes</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%&gt;%</span> 
  <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotSA.html">plotSA</a></span><span class="op">(</span>main<span class="op">=</span><span class="st">"Final model: Mean-variance trend"</span><span class="op">)</span></code></pre></div>
<p><img src="tidy_limmaWorkflow_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div id="useful-graphical-representations-of-differential-expression-results" class="section level2">
<h2 class="hasAnchor">
<a href="#useful-graphical-representations-of-differential-expression-results" class="anchor"></a>Useful graphical representations of differential expression results</h2>
<p>With <code>ggplot2</code> we We can reproduce and customise the plot for the association between fold-change and average log-abundance</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_test</span> <span class="op">%&gt;%</span>
  
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">.abundant</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Label significant</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>significant <span class="op">=</span> <span class="va">`adj.P.Val___groupBasal-groupLP`</span><span class="op">&lt;</span><span class="fl">0.05</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Subset labels</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/ifelse.html">ifelse</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span><span class="op">(</span><span class="va">`logFC___groupBasal-groupLP`</span><span class="op">)</span> <span class="op">&gt;=</span><span class="fl">8</span>, <span class="fu"><a href="https://rdrr.io/r/base/character.html">as.character</a></span><span class="op">(</span><span class="va">symbol</span><span class="op">)</span>, <span class="st">""</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
    x<span class="op">=</span><span class="va">`AveExpr___groupBasal-groupLP`</span>, 
    y<span class="op">=</span><span class="va">`logFC___groupBasal-groupLP`</span>, 
    label<span class="op">=</span><span class="va">symbol</span>
  <span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>color <span class="op">=</span> <span class="va">significant</span>, size <span class="op">=</span> <span class="va">significant</span>, alpha<span class="op">=</span><span class="va">significant</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  
  <span class="co"># Customisation</span>
  <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html">geom_text_repel</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#e11f28"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Using size for a discrete variable is not advised.</code></pre>
<pre><code>## Warning: Using alpha for a discrete variable is not advised.</code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C               LC_TIME=en_US.UTF-8       
##  [4] LC_COLLATE=en_US.UTF-8     LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                  LC_ADDRESS=C              
## [10] LC_TELEPHONE=C             LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
##  [1] tidybulk_1.2.0     ggrepel_0.8.2      plotly_4.9.2.1     ggplot2_3.3.2     
##  [5] magrittr_2.0.1     readr_1.4.0        tidyr_1.1.2        dplyr_1.0.2       
##  [9] tibble_3.0.4       zhejiang2020_0.3.3 BiocStyle_2.18.1  
## 
## loaded via a namespace (and not attached):
##  [1] bitops_1.0-6                matrixStats_0.57.0          fs_1.5.0                   
##  [4] bit64_4.0.5                 RColorBrewer_1.1-2          httr_1.4.2                 
##  [7] rprojroot_2.0.2             GenomeInfoDb_1.26.1         tools_4.0.3                
## [10] utf8_1.1.4                  R6_2.5.0                    DBI_1.1.0                  
## [13] lazyeval_0.2.2              BiocGenerics_0.36.0         colorspace_2.0-0           
## [16] withr_2.3.0                 tidyselect_1.1.0            gridExtra_2.3              
## [19] bit_4.0.4                   compiler_4.0.3              preprocessCore_1.52.0      
## [22] textshaping_0.2.1           cli_2.2.0                   Biobase_2.50.0             
## [25] desc_1.2.0                  DelayedArray_0.16.0         labeling_0.4.2             
## [28] bookdown_0.21               scales_1.1.1                dittoSeq_1.2.2             
## [31] ggridges_0.5.2              pkgdown_1.6.1               systemfonts_0.3.2          
## [34] stringr_1.4.0               digest_0.6.27               rmarkdown_2.5              
## [37] XVector_0.30.0              pkgconfig_2.0.3             htmltools_0.5.0            
## [40] MatrixGenerics_1.2.0        limma_3.46.0                htmlwidgets_1.5.2          
## [43] rlang_0.4.9                 RSQLite_2.2.1               farver_2.0.3               
## [46] generics_0.1.0              jsonlite_1.7.1              RCurl_1.98-1.2             
## [49] GenomeInfoDbData_1.2.4      Matrix_1.2-18               Rcpp_1.0.5                 
## [52] munsell_0.5.0               S4Vectors_0.28.0            fansi_0.4.1                
## [55] lifecycle_0.2.0             stringi_1.5.3               yaml_2.2.1                 
## [58] edgeR_3.32.0                SummarizedExperiment_1.20.0 zlibbioc_1.36.0            
## [61] plyr_1.8.6                  grid_4.0.3                  blob_1.2.1                 
## [64] parallel_4.0.3              crayon_1.3.4                lattice_0.20-41            
## [67] cowplot_1.1.0               hms_0.5.3                   locfit_1.5-9.4             
## [70] knitr_1.30                  pillar_1.4.7                GenomicRanges_1.42.0       
## [73] stats4_4.0.3                glue_1.4.2                  evaluate_0.14              
## [76] data.table_1.13.4           BiocManager_1.30.10         vctrs_0.3.5                
## [79] org.Mm.eg.db_3.12.0         gtable_0.3.0                purrr_0.3.4                
## [82] assertthat_0.2.1            xfun_0.19                   ragg_0.4.0                 
## [85] viridisLite_0.3.0           SingleCellExperiment_1.12.0 pheatmap_1.0.12            
## [88] AnnotationDbi_1.52.0        memoise_1.1.0               IRanges_2.24.0             
## [91] ellipsis_0.3.1</code></pre>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-robinson2010scaling">
<p>Robinson, Mark D, and Alicia Oshlack. 2010. “A Scaling Normalization Method for Differential Expression Analysis of Rna-Seq Data.” <em>Genome Biology</em> 11 (3): 1–9.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Xueyi Dong, Luyi Tian, Yue You, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
