<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Single-cell Tidy Transcriptomics - analysis of single-cell RNA sequencing data with R tidy principles • zhejiang2020</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Single-cell Tidy Transcriptomics - analysis of single-cell RNA sequencing data with R tidy principles">
<meta property="og:description" content="zhejiang2020">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--><!-- Global site tag (gtag.js) - Google Analytics --><script async src="https://www.googletagmanager.com/gtag/js?id=UA-93043521-1"></script><script>
  window.dataLayer = window.dataLayer || [];
  function gtag(){dataLayer.push(arguments);}
  gtag('js', new Date());

  gtag('config', 'UA-93043521-1');
</script>
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">zhejiang2020</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Bulk analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">R base</li>
    <li>
      <a href="../articles/limmaWorkflow.html">English</a>
    </li>
    <li>
      <a href="../articles/limmaWorkflow_chinese.html">Chinese</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Tidy transcriptomics</li>
    <li>
      <a href="../articles/tidy_limmaWorkflow.html">English</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Single-cell analyses
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">R base</li>
    <li>
      <a href="../articles/sc_analysis.html">Chinese</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Tidy transcriptomics</li>
    <li>
      <a href="../articles/tidy_single_Cell.html">English</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/stemangiola/zhejiang2020">
    <span class="fa fa-github"></span>
     
  </a>
</li>
<li>
  <a href="https://hub.docker.com/repository/docker/stemangiola/zhejiang2020">
    <span class="fa fa-docker"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="tidy_single_cell_files/accessible-code-block-0.0.1/empty-anchor.js"></script><link href="tidy_single_cell_files/anchor-sections-1.0/anchor-sections.css" rel="stylesheet">
<script src="tidy_single_cell_files/anchor-sections-1.0/anchor-sections.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Single-cell Tidy Transcriptomics - analysis of single-cell RNA sequencing data with R tidy principles</h1>
                        <h4 class="author">Stefano Mangiola</h4>
            <address class="author_afil">
      The Walter and Eliza Hall Institute of Medical Research, 1G Royal Parade, Parkville, VIC 3052, Melbourne, Australia; Department of Medical Biology, The University of Melbourne, Parkville, VIC 3010, Melbourne, Australia<br><h4 class="author">Maria Doyle</h4>
            <address class="author_afil">
      Peter MacCallum Cancer Centre, 305 Grattan Street, Parkville, Melbourne, Victoria, Australia<br><h4 class="date">11 November 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/stemangiola/zhejiang2020/blob/master/vignettes/tidy_single_cell.Rmd"><code>vignettes/tidy_single_cell.Rmd</code></a></small>
      <div class="hidden name"><code>tidy_single_cell.Rmd</code></div>

    </address>
</address>
</div>

    
    
<p><img src="../inst/vignettes/tidybulk_logo.png" width="100px"></p>
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
</div>
<div id="set-up" class="section level1">
<h1 class="hasAnchor">
<a href="#set-up" class="anchor"></a>Set-up</h1>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://stemangiola.github.io/zhejiang2020/">zhejiang2020</a></span><span class="op">)</span>

<span class="co"># Bioconductor</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/">scater</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">EnsDb.Hsapiens.v86</span><span class="op">)</span>

<span class="co"># tidyverse core packages</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tibble.tidyverse.org/">tibble</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyr.tidyverse.org">tidyr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">magrittr</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/eclarke/ggbeeswarm">ggbeeswarm</a></span><span class="op">)</span>

<span class="co">#library(tidyHeatmap)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidySingleCellExperiment</span><span class="op">)</span></code></pre></div>
<div id="participation" class="section level3">
<h3 class="hasAnchor">
<a href="#participation" class="anchor"></a>Participation</h3>
<p>After the lecture, participants are expected to follow along the hands-on session. we highly recommend participants bringing your own laptop.</p>
</div>
<div id="r-bioconductor-packages-used" class="section level3">
<h3 class="hasAnchor">
<a href="#r-bioconductor-packages-used" class="anchor"></a><em>R</em> / <em>Bioconductor</em> packages used</h3>
<p>The following R/Bioconductor packages will be explicitly used:</p>
<ul>
<li>tidySingleCellExperiment</li>
<li>DropletUtils</li>
<li>scran</li>
<li>scater</li>
<li>singleR</li>
</ul>
</div>
<div id="time-outline" class="section level3">
<h3 class="hasAnchor">
<a href="#time-outline" class="anchor"></a>Time outline</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Analysis workflow</td>
<td>1.5h</td>
</tr>
<tr class="even">
<td>Q &amp; A</td>
<td>30m</td>
</tr>
</tbody>
</table>
</div>
<div id="数据准备" class="section level2">
<h2 class="hasAnchor">
<a href="#%E6%95%B0%E6%8D%AE%E5%87%86%E5%A4%87" class="anchor"></a>数据准备</h2>
<p>We get the original SingleCellExperiment data used in the previous session.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">zhejiang2020</span><span class="fu">::</span><span class="va"><a href="../reference/single_cell_experiment.html">single_cell_experiment</a></span></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 33694 4290 
## metadata(1): Samples
## assays(1): counts
## rownames(33694): ENSG00000243485 ENSG00000237613 ... ENSG00000277475
##   ENSG00000268674
## rowData names(2): ID Symbol
## colnames(4290): AAACCTGAGAAGGCCT-1 AAACCTGAGACAGACC-1 ...
##   TTTGTCAGTTAAGACA-1 TTTGTCATCCCAAGAT-1
## colData names(2): Sample Barcode
## reducedDimNames(0):
## altExpNames(0):</code></pre>
<p>We can get a tidy representation where cell-wise information is displayed. The dataframe is displayed as <code>tibble abstraction</code>, to indicate that appears and act as a <code>tibble</code>, but underlies a <code>SingleCellExperiment</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts</span> <span class="op">=</span> 
  <span class="fu">zhejiang2020</span><span class="fu">::</span><span class="va"><a href="../reference/single_cell_experiment.html">single_cell_experiment</a></span> <span class="op">%&gt;%</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">counts</span></code></pre></div>
<pre><code>## # A tibble abstraction: 4,290 x 3
##    cell               Sample                      Barcode           
##    &lt;chr&gt;              &lt;chr&gt;                       &lt;chr&gt;             
##  1 AAACCTGAGAAGGCCT-1 raw_gene_bc_matrices/GRCh38 AAACCTGAGAAGGCCT-1
##  2 AAACCTGAGACAGACC-1 raw_gene_bc_matrices/GRCh38 AAACCTGAGACAGACC-1
##  3 AAACCTGAGATAGTCA-1 raw_gene_bc_matrices/GRCh38 AAACCTGAGATAGTCA-1
##  4 AAACCTGAGGCATGGT-1 raw_gene_bc_matrices/GRCh38 AAACCTGAGGCATGGT-1
##  5 AAACCTGCAAGGTTCT-1 raw_gene_bc_matrices/GRCh38 AAACCTGCAAGGTTCT-1
##  6 AAACCTGCAGGATTGG-1 raw_gene_bc_matrices/GRCh38 AAACCTGCAGGATTGG-1
##  7 AAACCTGCAGGCGATA-1 raw_gene_bc_matrices/GRCh38 AAACCTGCAGGCGATA-1
##  8 AAACCTGCATGAAGTA-1 raw_gene_bc_matrices/GRCh38 AAACCTGCATGAAGTA-1
##  9 AAACCTGGTAAATGAC-1 raw_gene_bc_matrices/GRCh38 AAACCTGGTAAATGAC-1
## 10 AAACCTGGTACATCCA-1 raw_gene_bc_matrices/GRCh38 AAACCTGGTACATCCA-1
## # … with 4,280 more rows</code></pre>
<p>If we need, we can extract transcript information too. A regular tibble will be returned for independent visualisation and analyses.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/join_transcripts.html">join_transcripts</a></span><span class="op">(</span><span class="st">"ENSG00000228463"</span><span class="op">)</span></code></pre></div>
<pre><code>## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</code></pre>
<pre><code>## # A tibble: 4,290 x 5
##    cell          transcript    abundance_counts Sample             Barcode      
##    &lt;chr&gt;         &lt;chr&gt;                    &lt;dbl&gt; &lt;chr&gt;              &lt;chr&gt;        
##  1 AAACCTGAGAAG… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGAGAAG…
##  2 AAACCTGAGACA… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGAGACA…
##  3 AAACCTGAGATA… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGAGATA…
##  4 AAACCTGAGGCA… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGAGGCA…
##  5 AAACCTGCAAGG… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGCAAGG…
##  6 AAACCTGCAGGA… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGCAGGA…
##  7 AAACCTGCAGGC… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGCAGGC…
##  8 AAACCTGCATGA… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGCATGA…
##  9 AAACCTGGTAAA… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGGTAAA…
## 10 AAACCTGGTACA… ENSG00000228…                0 raw_gene_bc_matri… AAACCTGGTACA…
## # … with 4,280 more rows</code></pre>
<p>对于得到的基因，根据gene ID 进行基因注释，得到基因名，并且知道其所在染色体（用于之后的细胞质量控制）。</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#--- gene-annotation ---#</span>
<span class="fu"><a href="https://tibble.tidyverse.org/reference/rownames.html">rownames</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span> <span class="op">&lt;-</span> 
  <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/uniquifyFeatureNames.html">uniquifyFeatureNames</a></span><span class="op">(</span>
    <span class="fu">rowData</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">$</span><span class="va">ID</span>,
    <span class="fu">rowData</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol</span>
  <span class="op">)</span></code></pre></div>
<p>这里，我们认为线粒体基因表达量高的barcode对应细胞质mRNA已经流出的破损细胞。并用这个指标来进行筛查过滤。</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Gene product location</span>
<span class="va">location</span> <span class="op">&lt;-</span> <span class="fu">mapIds</span><span class="op">(</span>
  <span class="va">EnsDb.Hsapiens.v86</span>, 
  keys<span class="op">=</span><span class="fu">rowData</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">$</span><span class="va">ID</span>, 
  column<span class="op">=</span><span class="st">"SEQNAME"</span>,
  keytype<span class="op">=</span><span class="st">"GENEID"</span>
<span class="op">)</span></code></pre></div>
<pre><code>## Warning: Unable to map 144 of 33694 requested IDs.</code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#--- quality-control ---#</span>
<span class="va">counts_annotated</span> <span class="op">=</span> 
  <span class="va">counts</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Join mitochondrion statistics</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate-joins.html">left_join</a></span><span class="op">(</span>
    <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/perCellQCMetrics.html">perCellQCMetrics</a></span><span class="op">(</span><span class="va">.</span>, subsets<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>Mito<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/which.html">which</a></span><span class="op">(</span><span class="va">location</span><span class="op">==</span><span class="st">"MT"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html">as_tibble</a></span><span class="op">(</span>rownames<span class="op">=</span><span class="st">"cell"</span><span class="op">)</span>,
    by<span class="op">=</span><span class="st">"cell"</span>
  <span class="op">)</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Label cells</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>high_mitochondrion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html">isOutlier</a></span><span class="op">(</span><span class="va">subsets_Mito_percent</span>, type<span class="op">=</span><span class="st">"higher"</span><span class="op">)</span><span class="op">)</span> </code></pre></div>
<p>We can plot various statistics</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_annotated</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotColData.html">plotColData</a></span><span class="op">(</span>
    y <span class="op">=</span> <span class="st">"subsets_Mito_percent"</span>,
    colour_by <span class="op">=</span> <span class="st">"high_mitochondrion"</span>
  <span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Mito percent"</span><span class="op">)</span></code></pre></div>
<p><img src="tidy_single_cell_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_annotated</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="fl">1</span>,y<span class="op">=</span><span class="va">subsets_Mito_percent</span>,
             color <span class="op">=</span> <span class="va">high_mitochondrion</span>, 
             alpha<span class="op">=</span><span class="va">high_mitochondrion</span>,
             size<span class="op">=</span> <span class="va">high_mitochondrion</span>
            <span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu">ggbeeswarm</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ggbeeswarm/man/geom_quasirandom.html">geom_quasirandom</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>

  <span class="co"># Customisation</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html">scale_color_manual</a></span><span class="op">(</span>values<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"#e11f28"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_size.html">scale_size_discrete</a></span><span class="op">(</span>range <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## Warning: Using size for a discrete variable is not advised.</code></pre>
<pre><code>## Warning: Using alpha for a discrete variable is not advised.</code></pre>
<p><img src="tidy_single_cell_files/figure-html/unnamed-chunk-7-2.png" width="700"></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_filtered</span> <span class="op">=</span> 
  <span class="va">counts_annotated</span> <span class="op">%&gt;%</span>
  
  <span class="co"># Filter data</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="op">!</span><span class="va">high_mitochondrion</span><span class="op">)</span>

<span class="va">counts_filtered</span></code></pre></div>
<pre><code>## # A tibble abstraction: 3,978 x 10
##    cell  Sample Barcode   sum detected subsets_Mito_sum subsets_Mito_de…
##    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;dbl&gt;    &lt;int&gt;            &lt;dbl&gt;            &lt;int&gt;
##  1 AAAC… raw_g… AAACCT…  1738      748              111               11
##  2 AAAC… raw_g… AAACCT…  3240     1052              177               12
##  3 AAAC… raw_g… AAACCT…  2983      951               67               11
##  4 AAAC… raw_g… AAACCT…  4181     1248               93               10
##  5 AAAC… raw_g… AAACCT… 12147     2785              394               11
##  6 AAAC… raw_g… AAACCT…  8588     2278              385               11
##  7 AAAC… raw_g… AAACCT…  3158     1118              109               11
##  8 AAAC… raw_g… AAACCT…  5081     1437               90               10
##  9 AAAC… raw_g… AAACCT…  3562      970               66               11
## 10 AAAC… raw_g… AAACCT…  7157     1639              173               12
## # … with 3,968 more rows, and 3 more variables: subsets_Mito_percent &lt;dbl&gt;,
## #   total &lt;dbl&gt;, high_mitochondrion &lt;lgl&gt;</code></pre>
<div id="scaling" class="section level3">
<h3 class="hasAnchor">
<a href="#scaling" class="anchor"></a>Scaling</h3>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#--- normalization ---#</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">)</span>

<span class="co"># Calculate clusters</span>
<span class="va">clusters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html">quickCluster</a></span><span class="op">(</span><span class="va">counts_filtered</span><span class="op">)</span>

<span class="co"># Add scaled counts</span>
<span class="va">counts_scaled</span> <span class="op">&lt;-</span> 
  <span class="va">counts_filtered</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html">computeSumFactors</a></span><span class="op">(</span>cluster<span class="op">=</span><span class="va">clusters</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html">logNormCounts</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/tidySingleCellExperiment/man/join_transcripts.html">join_transcripts</a></span><span class="op">(</span><span class="st">"CD79B"</span><span class="op">)</span></code></pre></div>
<pre><code>## tidySingleCellExperiment says: A data frame is returned for independent data analysis.</code></pre>
<pre><code>## # A tibble: 3,978 x 14
##    cell  transcript abundance_counts abundance_logco… Sample Barcode   sum
##    &lt;chr&gt; &lt;chr&gt;                 &lt;dbl&gt;            &lt;dbl&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;dbl&gt;
##  1 AAAC… CD79B                     0             0    raw_g… AAACCT…  1738
##  2 AAAC… CD79B                     1             1.21 raw_g… AAACCT…  3240
##  3 AAAC… CD79B                     0             0    raw_g… AAACCT…  2983
##  4 AAAC… CD79B                     0             0    raw_g… AAACCT…  4181
##  5 AAAC… CD79B                     0             0    raw_g… AAACCT… 12147
##  6 AAAC… CD79B                     6             2.05 raw_g… AAACCT…  8588
##  7 AAAC… CD79B                     4             2.75 raw_g… AAACCT…  3158
##  8 AAAC… CD79B                     0             0    raw_g… AAACCT…  5081
##  9 AAAC… CD79B                     0             0    raw_g… AAACCT…  3562
## 10 AAAC… CD79B                     0             0    raw_g… AAACCT…  7157
## # … with 3,968 more rows, and 7 more variables: detected &lt;int&gt;,
## #   subsets_Mito_sum &lt;dbl&gt;, subsets_Mito_detected &lt;int&gt;,
## #   subsets_Mito_percent &lt;dbl&gt;, total &lt;dbl&gt;, high_mitochondrion &lt;lgl&gt;,
## #   sizeFactor &lt;dbl&gt;</code></pre>
</div>
<div id="detect-variable-gene-transcripts" class="section level3">
<h3 class="hasAnchor">
<a href="#detect-variable-gene-transcripts" class="anchor"></a>Detect variable gene-transcripts</h3>
<p>为了减轻下游分析工具的计算负担，减少数据中的噪声，我们需要先进行特征选择。</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#--- variance-modelling ---#</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">1001</span><span class="op">)</span>
<span class="va">gene_variability</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVarByPoisson.html">modelGeneVarByPoisson</a></span><span class="op">(</span><span class="va">counts_scaled</span><span class="op">)</span>
<span class="va">top_variable</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html">getTopHVGs</a></span><span class="op">(</span><span class="va">gene_variability</span>, prop<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></code></pre></div>
</div>
<div id="dimensionality-reduction" class="section level3">
<h3 class="hasAnchor">
<a href="#dimensionality-reduction" class="anchor"></a>Dimensionality reduction</h3>
<p>特征选择后，单细胞表达矩阵的维数可以通过专门的降维算法进一步降低。使得数据可以直观可视化，并且将数据简化为基本组成部分。</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co">#--- dimensionality-reduction ---#</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">10000</span><span class="op">)</span>
<span class="va">counts_reduction</span> <span class="op">&lt;-</span> 
  <span class="va">counts_scaled</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scran/man/denoisePCA.html">denoisePCA</a></span><span class="op">(</span>subset.row<span class="op">=</span><span class="va">top_variable</span>, technical<span class="op">=</span><span class="va">gene_variability</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runTSNE.html">runTSNE</a></span><span class="op">(</span>dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html">runUMAP</a></span><span class="op">(</span>dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span>

<span class="va">counts_reduction</span></code></pre></div>
<pre><code>## # A tibble abstraction: 3,978 x 20
##    cell  Sample Barcode   sum detected subsets_Mito_sum subsets_Mito_de…
##    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;dbl&gt;    &lt;int&gt;            &lt;dbl&gt;            &lt;int&gt;
##  1 AAAC… raw_g… AAACCT…  1738      748              111               11
##  2 AAAC… raw_g… AAACCT…  3240     1052              177               12
##  3 AAAC… raw_g… AAACCT…  2983      951               67               11
##  4 AAAC… raw_g… AAACCT…  4181     1248               93               10
##  5 AAAC… raw_g… AAACCT… 12147     2785              394               11
##  6 AAAC… raw_g… AAACCT…  8588     2278              385               11
##  7 AAAC… raw_g… AAACCT…  3158     1118              109               11
##  8 AAAC… raw_g… AAACCT…  5081     1437               90               10
##  9 AAAC… raw_g… AAACCT…  3562      970               66               11
## 10 AAAC… raw_g… AAACCT…  7157     1639              173               12
## # … with 3,968 more rows, and 13 more variables: subsets_Mito_percent &lt;dbl&gt;,
## #   total &lt;dbl&gt;, high_mitochondrion &lt;lgl&gt;, sizeFactor &lt;dbl&gt;, PC1 &lt;dbl&gt;,
## #   PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;, TSNE1 &lt;dbl&gt;, TSNE2 &lt;dbl&gt;,
## #   UMAP1 &lt;dbl&gt;, UMAP2 &lt;dbl&gt;</code></pre>
</div>
</div>
<div id="clustering" class="section level2">
<h2 class="hasAnchor">
<a href="#clustering" class="anchor"></a>Clustering</h2>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_cluster</span> <span class="op">&lt;-</span>
  <span class="va">counts_reduction</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>
    cluster <span class="op">=</span> 
      <span class="fu"><a href="https://rdrr.io/pkg/scran/man/buildSNNGraph.html">buildSNNGraph</a></span><span class="op">(</span><span class="va">.</span>, k<span class="op">=</span><span class="fl">10</span>, use.dimred <span class="op">=</span> <span class="st">'PCA'</span><span class="op">)</span> <span class="op">%&gt;%</span>
      <span class="fu">igraph</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/igraph/man/cluster_louvain.html">cluster_louvain</a></span><span class="op">(</span><span class="op">)</span> <span class="op">%$%</span>
      <span class="va">membership</span> <span class="op">%&gt;%</span>
      <span class="fu"><a href="https://rdrr.io/r/base/factor.html">as.factor</a></span><span class="op">(</span><span class="op">)</span>
  <span class="op">)</span>
  
<span class="va">counts_cluster</span></code></pre></div>
<pre><code>## # A tibble abstraction: 3,978 x 21
##    cell  Sample Barcode   sum detected subsets_Mito_sum subsets_Mito_de…
##    &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;   &lt;dbl&gt;    &lt;int&gt;            &lt;dbl&gt;            &lt;int&gt;
##  1 AAAC… raw_g… AAACCT…  1738      748              111               11
##  2 AAAC… raw_g… AAACCT…  3240     1052              177               12
##  3 AAAC… raw_g… AAACCT…  2983      951               67               11
##  4 AAAC… raw_g… AAACCT…  4181     1248               93               10
##  5 AAAC… raw_g… AAACCT… 12147     2785              394               11
##  6 AAAC… raw_g… AAACCT…  8588     2278              385               11
##  7 AAAC… raw_g… AAACCT…  3158     1118              109               11
##  8 AAAC… raw_g… AAACCT…  5081     1437               90               10
##  9 AAAC… raw_g… AAACCT…  3562      970               66               11
## 10 AAAC… raw_g… AAACCT…  7157     1639              173               12
## # … with 3,968 more rows, and 14 more variables: subsets_Mito_percent &lt;dbl&gt;,
## #   total &lt;dbl&gt;, high_mitochondrion &lt;lgl&gt;, sizeFactor &lt;dbl&gt;, cluster &lt;fct&gt;,
## #   PC1 &lt;dbl&gt;, PC2 &lt;dbl&gt;, PC3 &lt;dbl&gt;, PC4 &lt;dbl&gt;, PC5 &lt;dbl&gt;, TSNE1 &lt;dbl&gt;,
## #   TSNE2 &lt;dbl&gt;, UMAP1 &lt;dbl&gt;, UMAP2 &lt;dbl&gt;</code></pre>
<p>使用tSNE对聚类进行可视化。</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plot_reddim.html">plotTSNE</a></span><span class="op">(</span><span class="va">counts_cluster</span>, colour_by<span class="op">=</span><span class="st">"cluster"</span>,text_by<span class="op">=</span><span class="st">"cluster"</span><span class="op">)</span></code></pre></div>
<p><img src="tidy_single_cell_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">counts_cluster</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>
    <span class="va">TSNE1</span>, <span class="va">TSNE2</span>, 
    color<span class="op">=</span><span class="va">cluster</span>, 
    size <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="va">subsets_Mito_percent</span> 
  <span class="op">)</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span>alpha<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><img src="tidy_single_cell_files/figure-html/unnamed-chunk-13-2.png" width="700"></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Xueyi Dong, Luyi Tian, Yue You, Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
